### PR TITLE
Support customized validation failure handler

### PR DESCRIPTION
**Pull request by springuper**
It's a strong requirement to show validation error msg to user, that's why I create this PR.

### CODE REVIEW

File: src/RIEStatefulBase.js [hunk 1]
```diff
@@ -13,11 +13,15 @@ export default class RIEStatefulBase extends RIEBase {
...
     finishEditing = () => {
         let newValue = ReactDOM.findDOMNode(this.refs.input).value;
-        this.doValidations(newValue);
-        if(!this.state.invalid && this.props.value !== newValue) {
+        const result = this.doValidations(newValue);
+        if(result && this.props.value !== newValue) {
             this.commit(newValue);
         }
-        this.cancelEditing();
+        if(!result && this.props.handleValidationFail) {
```
Comment (springuper) [inline src/RIEStatefulBase.js]: ideally, `this.doValidation` should return `false` when no error and `error message` when validation fails. But that will change current `validate` interface.

a workaround is, given the `newValue`, re-validate it and get the error message.

File: src/RIEBase.js [hunk 2]
```diff
@@ -33,11 +34,15 @@ export default class RIEBase extends React.Component {
...
         if(this.props.validate) {
-            this.setState({invalid: !this.props.validate(value)});
+            result = this.props.validate(value);
         } else if (this.validate) {
-            this.setState({invalid: !this.validate(value)});
+            result = this.validate(value);
         }
+        this.setState({invalid: !result});
+
+        return result;
```
Comment (springuper) [inline src/RIEBase.js]: react's `setState` is async, so you can not get the new state in the following sync logics. change it to a local variable and return it directly.

File: src/RIEBase.js [hunk 3]
```diff
@@ -49,7 +55,9 @@ export default class RIEBase extends React.Component {
     };
 
     componentWillReceiveProps = (nextProps) => {
-        if ('value' in nextProps) this.setState({loading: false, editing: false, invalid: false, newValue: null});
+        if ('value' in nextProps && !(nextProps.shouldRemainWhileInvalid && this.state.invalid)) {
```
Comment (springuper) [inline src/RIEBase.js]: remain editing mode when new value is invalid

Comment (elihofni) [reply src/RIEBase.js:58]: Not working with me

**Pull request merged**